rust/core/src/driver_manager.rs

```diff
@@ -0,0 +1,1218 @@
+#[derive(Clone)]
+pub struct ManagedDriver {
+    inner: Arc<DriverManagerInner>,
+}
+
+impl ManagedDriver {
+    /// Load a driver from an initialization function.
+    pub fn load_static(init: &ffi::FFI_AdbcDriverInitFunc, version: AdbcVersion) -> Result<Self> {
+        let driver = Self::load_impl(init, version)?;
+        let inner = Arc::new(DriverManagerInner {
+            driver: Mutex::new(driver),
+            version,
+            _library: None,
+        });
+        Ok(ManagedDriver { inner })
+    }
+
+    /// Load a driver from a dynamic library.
+    ///
+    /// Will attempt to load the dynamic library with the given `name`, find the
+    /// symbol with name `entrypoint` (defaults to `AdbcDriverInit` if `None`),
+    /// and then call create the driver using the resolved function.
+    ///
+    /// The `name` should not include any platform-specific prefixes or suffixes.
+    /// For example, use `adbc_driver_sqlite` rather than `libadbc_driver_sqlite.so`.
+    pub fn load_dynamic(
+        name: &str,
+        entrypoint: Option<&[u8]>,
+        version: AdbcVersion,
+    ) -> Result<Self> {
+        let entrypoint = entrypoint.unwrap_or(b"AdbcDriverInit");
+        let library = unsafe { libloading::Library::new(libloading::library_filename(name))? };
+        let init: libloading::Symbol<ffi::FFI_AdbcDriverInitFunc> =
+            unsafe { library.get(entrypoint)? };
+        let driver = Self::load_impl(&init, version)?;
+        let inner = Arc::new(DriverManagerInner {
+            driver: Mutex::new(driver),
+            version,
+            _library: Some(library),
+        });
+        Ok(ManagedDriver { inner })
+    }
+
+    fn load_impl(
+        init: &ffi::FFI_AdbcDriverInitFunc,
+        version: AdbcVersion,
+    ) -> Result<ffi::FFI_AdbcDriver> {
+        let mut error = ffi::FFI_AdbcError::default();
+        let mut driver = ffi::FFI_AdbcDriver::default();
+        let status = unsafe {
+            init(
+                version.into(),
+                &mut driver as *mut ffi::FFI_AdbcDriver as *mut c_void,
+                &mut error,
+            )
+        };
+        check_status(status, error)?;
+        Ok(driver)
+    }
+}
+
+impl Driver for ManagedDriver {
+    type DatabaseType = ManagedDatabase;
+
+    fn new_database(&mut self) -> Result<Self::DatabaseType> {
+        self.new_database_with_opts([])
+    }
+
+    fn new_database_with_opts(
+        &mut self,
+        opts: impl IntoIterator<Item = (<Self::DatabaseType as Optionable>::Option, OptionValue)>,
+    ) -> Result<Self::DatabaseType> {
+        let driver = &self.inner.driver.lock().unwrap();
+        let mut database = ffi::FFI_AdbcDatabase::default();
+
+        // DatabaseNew
+        let mut error = ffi::FFI_AdbcError::with_driver(driver);
+        let method = driver_method!(driver, DatabaseNew);
+        let status = unsafe { method(&mut database, &mut error) };
+        check_status(status, error)?;
```

Comment:
I think that this will made the semantics of `check_status` more complex without any real benefit: the pattern of reusing the error can only be seen in `new_database_with_opts` and `new_connection_with_opts`. It seems like a micro-optimization to me (and those errors are probably stack-allocated anyway).